From 040ee1e57aff3dcb2f92904e918fd05e9e15ae00 Mon Sep 17 00:00:00 2001
From: Don Zickus <dzickus@redhat.com>
Date: Fri, 12 Sep 2014 18:03:56 -0400
Subject: [powerpc] Fix string instr. emulation for 32-bit processes on ppc64

Message-id: <1410545655-205645-8-git-send-email-dzickus@redhat.com>
Patchwork-id: 94095
O-Subject: [RHEL7 PATCH 007/626] powerpc: Fix string instr. emulation for 32-bit processes on ppc64
Bugzilla: 1127366
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>
RH-Acked-by: Steve Best <sbest@redhat.com>
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: David Gibson <dgibson@redhat.com>

https://bugzilla.redhat.com/show_bug.cgi?id=1127366

commit 80aa0fb4940bf8ee52bcb574d74459a7aea45621
Author: James Yang <James.Yang@freescale.com>
Date:   Tue Jun 25 11:41:05 2013 -0500

    powerpc: Fix string instr. emulation for 32-bit processes on ppc64

    String instruction emulation would erroneously result in a segfault if
    the upper bits of the EA are set and is so high that it fails access
    check.  Truncate the EA to 32 bits if the process is 32-bit.

    Signed-off-by: James Yang <James.Yang@freescale.com>
    Signed-off-by: Benjamin Herrenschmidt <benh@kernel.crashing.org>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/arch/powerpc/kernel/traps.c b/arch/powerpc/kernel/traps.c
index 3fc191a..64a1073 100644
--- a/arch/powerpc/kernel/traps.c
+++ b/arch/powerpc/kernel/traps.c
@@ -864,6 +864,10 @@ static int emulate_string_inst(struct pt_regs *regs, u32 instword)
   u8 val;
   u32 shift = 8 * (3 - (pos & 0x3));
 
+  /* if process is 32-bit, clear upper 32 bits of EA */
+  if ((regs->msr & MSR_64BIT) == 0)
+   EA &= 0xFFFFFFFF;
+
   switch ((instword & PPC_INST_STRING_MASK)) {
    case PPC_INST_LSWX:
    case PPC_INST_LSWI:
-- 
1.7.1